File: backend/app/services/task_service.py

from app.repositories.task_repository import TaskRepository, Task
from app.services.drone_assignment_service import DroneAssignmentService, DroneAssignment
from app.services.drone_image_service import DroneImage, DroneImageService


class TaskService:
    @staticmethod
    def get_all():
        return TaskRepository.get_all_tasks()

    @staticmethod
    def paginate(page: int = 1, per_page: int = 20):
        return TaskRepository.paginate(page=page, per_page=per_page)

    @staticmethod
    def get_task_by_id(task_id):
        return TaskRepository.get_task_by_id(task_id)

    @staticmethod
    def create_task(data):
        name = data.get("name", "")
        description = data.get("description")
        task = Task(name=name, description=description)
        TaskRepository.add_task(task)
        return task

    @staticmethod
    def update_task(task_id, data):
        task: Task = TaskRepository.get_task_by_id(task_id)
        if task:
            task.name = data.get('name', task.name)
            task.description = data.get('description', task.description)
            TaskRepository.update_task(task)
        return task

    @staticmethod
    def delete_task(task_id):
        task = TaskRepository.get_task_by_id(task_id)
        if task:
            TaskRepository.delete_task(task)
        return task

    @staticmethod
    def assign_drone(task_id: int, drone_id: int) -> DroneAssignment:
        assignment = DroneAssignmentService.find_by_drone_id_and_task_id(
            task_id=task_id, drone_id=drone_id)

        if (assignment):
            return assignment

        return DroneAssignmentService.create({"drone_id": drone_id, "task_id": task_id})

    @staticmethod
    def get_assignments(task_id):
        return DroneAssignmentService.filter_by_task(task_id=task_id)

    @staticmethod
    def get_images(task_id: int) -> list[DroneImage]:
        return DroneImageService.filter_by_task_id(task_id=task_id)

File: frontend/src/services/task_service.ts




import { Assignment } from "@/models/assignment.model";
import { Image } from "@/models/image.model";
import { Task } from "@/models/task.model";
import service from "@/services/droneqube_service";

type TaskCreateData = {
    name: string,
    description: string,
}

type AssignDroneData = {
    drone_id: number
}
class TaskService {
    static async getTasks(pageIndex:number=1) {
        const response = await service.get("/tasks", { params: { page: pageIndex } });
        if (response.data) {
            return response.data as Task[];
        }
    }
    static async createTask(_data: TaskCreateData) {
        const response = await service.post("/tasks", { ..._data });
        if (response.data) {
            return response.data as Task;
        }
    }
    static async assignDrone(taskId: number, _data: AssignDroneData) {
        const response = await service.post(`/tasks/${taskId}/assign-drone`, { ..._data });
        if (response.data) {
            return response.data as Assignment
        }
    }
    static async listImages(taskId: number) {
        const response = await service.get(`/tasks/${taskId}/images`);
        if (response.data) {
            return response.data as Image[]
        }
    }

}
export default TaskService;